Explosion config Overhaul

Quarri6343 opened the pull request:
**What:**
This PR
・fix inappropriate doExplosion config description
・add a new config option called doEnvironmentExplosion  that allows machines set on Fire and explode in Rain, Fire, Water or Lava

**Implementation Details:**
・I edited configHolder
・I added a new method called checkEnvironmentExplosion in TieredMetaTileEntity and MetaTileEntityEnergyHatch class to allow machines explode in certain environment
Its explosion condition is from GT5u
It won't be called if the explosion config is off
・I added a new method called getEnvironmentResistance
it should return true if the machine has resistance to certain environment
・I added a informative tooltip for some machines which have Environment Resistance
・I added a new argument in the existing doExplosion method to allow environment explosion can also be handled by it

**Outcome:**
fixes #958
gives more challenges to experienced GT players

**Additional info:**
Multiblocks Controller itself does not explode unless its Energy Hatch explodes

**Possible compatibility issue:**
If any addon implements solar panels, they must override getEnvironmentResistance in their class to prevent explosion when the config is enabled

ALongStringOfNumbers commented on src/main/java/gregtech/api/metatileentity/TieredMetaTileEntity.java:
```diff
@@ -68,6 +86,47 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
+            if (world.rand.nextInt(1000) == 0) {
+                if (world.isRainingAt(getPos()) || world.isRainingAt(getPos().east()) || world.isRainingAt(getPos().west()) || world.isRainingAt(getPos().north()) || world.isRainingAt(getPos().south())) {
+                    if (world.isThundering() && world.rand.nextInt(3) == 0) {
+                        doExplosion(tier, true);
```
Hm, I think that the terrain damage here should still respect the config option `doExplosion`. All the config option does is determine if the terrain is damaged when the machine explodes, it will explode on either config option.

ALongStringOfNumbers commented on src/main/java/gregtech/common/ConfigHolder.java:
```diff
@@ -74,9 +74,12 @@ public static class MachineOptions {
         public double multiblockSteamToEU = 0.5;
 
-        @Config.Comment({"Whether machines should explode when overloaded with power.", "Default: true"})
+        @Config.Comment({"Whether machines or boilers should explode when overloaded with power or met special conditions.", "Default: true"})
```
Could you change this config option to say "... explode causing terrain damage when overloaded ..."

Quarri6343 commented:
Thanks for the review!
`All the config option does is determine if the terrain is damaged when the machine explodes`
I changed the config name to doExplosionDamagesTerrain to avoid future confusion
`Could you change this config option to say "... explode causing terrain damage when overloaded ..."`
I did a slightly different approach to fix the description but it should tell the same thing

I also changed _Environment_ to _Terrain_
Its my bad

ALongStringOfNumbers commented on src/main/java/gregtech/api/metatileentity/MetaTileEntity.java:
```diff
@@ -1360,7 +1361,16 @@ public RecipeMap<?> getRecipeMap() {
+        for (EnumFacing side : EnumFacing.VALUES) {
+            if (getWorld().isAirBlock(getPos().offset(side))) {
+                getWorld().setBlockState(getPos().offset(side), Blocks.FIRE.getDefaultState(), 11);
+                return;
```
Instead of calling `return` here, which will result in only 1 air space being turned to fire, this method could take in a double for percentage chance to turn blocks to fire. It could look something like

```java
public void setOnFire(double chance) {
        for(EnumFacing side : EnumFacing.VALUES) {
        if(chance >= GTValues.RNG.nextDouble(100.00) && getWorld.isAirBlock(getPos().offset(side))) {
        getWorld().setBlockState(getPos().offset(side), Blocks.FIRE.getDefaultState(), 11);
        }
    }
}
```

This would allow for higher tier machines a chance for setting more sides on fire at one time, if the passed chance is based on the tier of the machine.

However, this does not guarantee at least one fire spawned, so you could also change the above method and have the passed `chance` value be the chance for an additional fire to be spawned after the first fire is spawned. (I think this makes the most sense)

ALongStringOfNumbers commented on src/main/java/gregtech/api/metatileentity/TieredMetaTileEntity.java:
```diff
@@ -68,6 +86,47 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
+     *
+     * @return true if tile entity should not explode in these sources
+     */
+    public boolean getTerrainResistance() {
```
This could be called something like `getIsWeatherOrTerrainResistant`

ALongStringOfNumbers commented on src/main/java/gregtech/api/metatileentity/TieredMetaTileEntity.java:
```diff
@@ -68,6 +86,47 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
+        }
+    }
+
+    public void checkTerrainExplosion() {
```
I would recommend `checkWeatherOrTerrainExplosion`

ALongStringOfNumbers commented on src/main/java/gregtech/common/ConfigHolder.java:
```diff
@@ -74,8 +74,12 @@ public static class MachineOptions {
-        @Config.Comment({"Whether machines should explode when overloaded with power.", "Default: true"})
-        public boolean doExplosions = true;
+        @Config.Comment({"Whether machines or boilers explosion damages terrain.",
+                "Note machines or boilers always explode when overloaded with power or met special conditions regardless this config.", "Default: true"})
```
`machines and boilers` and also `or met with special conditions, regardess of....`

ALongStringOfNumbers commented on src/main/java/gregtech/common/ConfigHolder.java:
```diff
@@ -74,8 +74,12 @@ public static class MachineOptions {
 
-        @Config.Comment({"Whether machines should explode when overloaded with power.", "Default: true"})
-        public boolean doExplosions = true;
+        @Config.Comment({"Whether machines or boilers explosion damages terrain.",
```
"Whether machines or boilers damage the terrain when they explode"

ALongStringOfNumbers commented on src/main/java/gregtech/common/ConfigHolder.java:
```diff
@@ -74,8 +74,12 @@ public static class MachineOptions {
+                "Note machines or boilers always explode when overloaded with power or met special conditions regardless this config.", "Default: true"})
+        public boolean doExplosionDamagesTerrain = true;
+
+        @Config.Comment({"Whether machines explode in terrain damage such as rain or fire.", "Default: false"})
```
"Whether machines explode in rainy weather or when placed next to certain terrain, such as fire or lava"

ALongStringOfNumbers commented on src/main/java/gregtech/api/metatileentity/TieredMetaTileEntity.java:
```diff
@@ -68,6 +86,47 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
+            if (world.rand.nextInt(1000) == 0) {
+                for (EnumFacing side : EnumFacing.VALUES) {
+                    Block block = getWorld().getBlockState(getPos().offset(side)).getBlock();
+                    if (block == Blocks.FIRE || block == Blocks.WATER || block == Blocks.FLOWING_WATER || block == Blocks.LAVA || block == Blocks.FLOWING_LAVA) {
```
Not sure if water should be included here, some discussion might be necessary on this one.

ALongStringOfNumbers commented on src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityRockBreaker.java:
```diff
@@ -101,4 +101,9 @@ protected boolean shouldSearchForRecipes() {
+
+    @Override
+    public boolean getTerrainResistance(){
+        return true;
```
This machine should also have the tooltip applied to it if the weather explosion config is true.

Quarri6343 replied:
TieredMTE class has this code so this machine already has the tooltip
```
    @Override
    public void addInformation(ItemStack stack, @Nullable World player, @Nonnull List<String> tooltip, boolean advanced) {
        super.addInformation(stack, player, tooltip, advanced);
        if (ConfigHolder.machines.doTerrainExplosion && getIsWeatherOrTerrainResistant())
            tooltip.add(I18n.format("gregtech.universal.tooltip.terrain_resist"));
    }
```

ALongStringOfNumbers commented on src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityEnergyHatch.java:
```diff
@@ -154,4 +166,37 @@ public void getSubItems(CreativeTabs creativeTab, NonNullList<ItemStack> subItem
         }
     }
+
+    public void checkTerrainExplosion() {
```
This method could be put in GTUtility or in MTE, so you don't have to copy and paste it for both the tiered MTEs and the energy hatches.

DStrand1 commented on src/main/java/gregtech/api/metatileentity/MetaTileEntity.java:
```diff
@@ -1357,10 +1359,59 @@ public RecipeMap<?> getRecipeMap() {
         return null;
     }
 
+    public void checkWeatherOrTerrainExplosion(float explosionPower, double additionalFireChance, IEnergyContainer energyContainer) {
```
I think it would be good to check the config here as well to avoid an accidental call and exit early

DStrand1 commented on src/main/java/gregtech/api/metatileentity/TieredMetaTileEntity.java:
```diff
@@ -68,6 +86,14 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
+    @Override
+    public void update() {
+        super.update();
+        if (ConfigHolder.machines.doTerrainExplosion) {
```
If we have the config check in MTE, we can get rid of the check here

DStrand1 commented on src/main/java/gregtech/common/ConfigHolder.java:
```diff
@@ -74,8 +74,12 @@ public static class MachineOptions {
-        public boolean doExplosions = true;
+        @Config.Comment({"Whether machines or boilers damage the terrain when they explode.",
+                "Note machines and boilers always explode when overloaded with power or met with special conditions, regardless of this config.", "Default: true"})
+        public boolean doExplosionDamagesTerrain = true;
```
`doesExplosionDamageTerrain` ?

DStrand1 commented on src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityFisher.java:
```diff
@@ -60,7 +61,7 @@ protected ModularUI createUI(EntityPlayer entityPlayer) {
 
         ModularUI.Builder builder = ModularUI.builder(GuiTextures.BACKGROUND, 176,
-                18 + 18 * rowSize + 94)
+                        18 + 18 * rowSize + 94)
```
Can you revert this, I think it may conflict with the ModularUI work and is a simple merge conflict to avoid

DStrand1 commented on src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityEnergyHatch.java:
```diff
@@ -58,6 +59,14 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
+    @Override
+    public void update() {
+        super.update();
+        if (ConfigHolder.machines.doTerrainExplosion) {
```
Can remove this config check as well if we check in MTE

Quarri6343 commented:
this will close #879 since I changed config description
